nullaway/src/test/java/com/uber/nullaway/VarargsTests.java

```diff
@@ -177,4 +207,211 @@ public void testSkipJetbrainsNotNullOnVarArgsFromThirdPartyJars() {
+            "    Utilities.takesNullableVarargsArray(o1, (java.lang.Object) null);",
+            "    // this is fine!",
+            "    Utilities.takesNullableVarargsArray(o1, (java.lang.Object[]) null);",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void typeUseAndDeclarationOnElements() {
+    defaultCompilationHelper
+        .addSourceLines(
+            "Nullable.java",
+            "package com.uber;",
+            "import java.lang.annotation.ElementType;",
+            "import java.lang.annotation.Target;",
+            "@Target({ElementType.METHOD, ElementType.FIELD, ElementType.PARAMETER, ElementType.LOCAL_VARIABLE, ElementType.TYPE_USE})",
+            "public @interface Nullable {}")
+        .addSourceLines(
+            "Utilities.java",
+            "package com.uber;",
+            "public class Utilities {",
+            " public static String takesNullableVarargsArray(Object o, @Nullable Object... others) {",
+            "  String s = o.toString() + \" \";",
+            "  for (Object other : others) {",
+            "    s += (other == null) ? \"(null) \" : other.toString() + \" \";",
+            "  }",
+            "  return s;",
+            " }",
+            "}")
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "public class Test {",
+            "  public void testNullableVarargsArray(Object o1, Object o2, Object o3, @Nullable Object o4) {",
+            "    Utilities.takesNullableVarargsArray(o1, o2, o3, o4);",
+            "    Utilities.takesNullableVarargsArray(o1);", // Empty var args passed
+            "    Utilities.takesNullableVarargsArray(o1, (java.lang.Object) null);",
+            "    // BUG: Diagnostic contains: passing @Nullable parameter '(java.lang.Object[]) null' where @NonNull",
+            "    Utilities.takesNullableVarargsArray(o1, (java.lang.Object[]) null);",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void typeUseAndDeclarationOnBoth() {
+    defaultCompilationHelper
+        .addSourceLines(
+            "Nullable.java",
+            "package com.uber;",
+            "import java.lang.annotation.ElementType;",
+            "import java.lang.annotation.Target;",
+            "@Target({ElementType.METHOD, ElementType.FIELD, ElementType.PARAMETER, ElementType.LOCAL_VARIABLE, ElementType.TYPE_USE})",
+            "public @interface Nullable {}")
+        .addSourceLines(
+            "Utilities.java",
+            "package com.uber;",
+            "public class Utilities {",
+            " public static String takesNullableVarargsArray(Object o, @Nullable Object @Nullable... others) {",
+            "  String s = o.toString() + \" \";",
+            "  // BUG: Diagnostic contains: enhanced-for expression others is @Nullable",
+            "  for (Object other : others) {",
+            "    s += (other == null) ? \"(null) \" : other.toString() + \" \";",
```

Comment:
Should some part of this test (and similar ones) check that `other` is `@Nullable` here? (i.e. that a deref without checking is an error)

Reply:
That reasoning only occurs in JSpecify mode.  It is tested in `JSpecifyVarargsTests#typeUseAndDeclarationOnBoth` and in other methods in that class.